File: src/console/command/impl/group/GroupCommand.php
<?php

namespace pocketcloud\cloud\console\command\impl\group;

use pocketcloud\cloud\console\command\Command;
use pocketcloud\cloud\console\command\parameter\def\ServerGroupParameter;
use pocketcloud\cloud\console\command\parameter\def\StringParameter;
use pocketcloud\cloud\console\command\sender\ICommandSender;
use pocketcloud\cloud\console\command\SubCommand;
use pocketcloud\cloud\group\ServerGroup;
use pocketcloud\cloud\group\ServerGroupManager;
use pocketcloud\cloud\template\TemplateManager;

final class GroupCommand extends Command {

    public function __construct() {
        parent::__construct("group", "Manage server groups", ["g", "sg", "servergroup"]);

        $this->registerSubCommand(SubCommand::fromClosure("create", $this->handleCreateSub(...))
            ->addParameter(new StringParameter("name", false))
            ->addParameter(new StringParameter("templates", true, true)));

        $this->registerSubCommand(SubCommand::fromClosure("remove", $this->handleRemoveSub(...))
            ->addParameter(new ServerGroupParameter("group", false)));

        $this->registerSubCommand(SubCommand::fromClosure("addt", $this->handleAddTemplateSub(...))
            ->addParameter(new ServerGroupParameter("group", false))
            ->addParameter(new StringParameter("templates", false, true)));

        $this->registerSubCommand(SubCommand::fromClosure("removet", $this->handleRemoveTemplateSub(...))
            ->addParameter(new ServerGroupParameter("group", false))
            ->addParameter(new StringParameter("templates", false, true)));

        $this->registerSubCommand(SubCommand::fromClosure("list", $this->handleListSub(...)));
    }

    public function run(ICommandSender $sender, string $label, array $args, ?SubCommand $subCommand = null): bool {
        return true;
    }

    private function handleCreateSub(ICommandSender $sender, string $label, array $args): bool {
        $name = $args["name"];
        if (ServerGroupManager::getInstance()->check($name)) {
            $sender->warn("A server group with that name already exists.");
            return true;
        }

        $templates = $args["templates"] ?? [];
        if (is_string($templates)) {
            $templates = explode(" ", $templates);
            foreach ($templates as $i => $template) {
                if (!TemplateManager::getInstance()->check($template)) unset($templates[$i]);
            }
        }

        ServerGroupManager::getInstance()->create(new ServerGroup($name, array_values($templates)));
        return true;
    }

    private function handleRemoveSub(ICommandSender $sender, string $label, array $args): bool {
        ServerGroupManager::getInstance()->remove($args["group"]);
        return true;
    }

    private function handleAddTemplateSub(ICommandSender $sender, string $label, array $args): bool {
        /** @var ServerGroup $group */
        $group = $args["group"];
        $templates = explode(" ", $args["templates"]);
        foreach ($templates as $template) {
            if (!TemplateManager::getInstance()->check($template)) {
                $sender->warn("A template with the name §b{} §rdoesn't exist.", $template);
                continue;
            }

            ServerGroupManager::getInstance()->addTemplate($group, TemplateManager::getInstance()->get($template));
        }

        return true;
    }

    private function handleRemoveTemplateSub(ICommandSender $sender, string $label, array $args): bool {
        /** @var ServerGroup $group */
        $group = $args["group"];
        $templates = explode(" ", $args["templates"]);
        foreach ($templates as $template) {
            if (!TemplateManager::getInstance()->check($template)) {
                $sender->warn("A template with the name §b{} §rdoesn't exist.", $template);
                continue;
            }

            ServerGroupManager::getInstance()->removeTemplate($group, TemplateManager::getInstance()->get($template));
        }

        return true;
    }

    private function handleListSub(ICommandSender $sender, string $label, array $args): bool {
        $sender->info("ServerGroups §8(§b" . count($groups = ServerGroupManager::getInstance()->getAll()) . "§8)§r:");
        if (empty($groups)) $sender->info("§cNo server groups found.");
        foreach ($groups as $group) {
            $sender->info("§b{} §8| §rPlayerCount: §b{} §8| §r§rTemplates: §b{}", $group->getName(), $group->getPlayerCount(), empty($group->getTemplates()) ? "§cNone" : implode("§8, §b", $group->getTemplates()));
        }
        return true;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/impl/group/GroupCommand.java
package de.pocketcloud.cloud.console.command.impl.group;

import de.pocketcloud.api.component.group.IServerGroup;
import de.pocketcloud.api.component.server.ICloudServer;
import de.pocketcloud.api.component.template.ITemplate;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.builder.ServerGroupBuilder;
import de.pocketcloud.cloud.console.command.Command;
import de.pocketcloud.cloud.console.command.ctx.CommandContext;
import de.pocketcloud.cloud.console.command.desc.CommandDescription;
import de.pocketcloud.cloud.console.command.parameter.def.*;
import de.pocketcloud.cloud.console.command.sender.CommandSender;
import de.pocketcloud.cloud.console.command.sub.SubCommand;
import de.pocketcloud.cloud.template.group.ServerGroup;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@CommandDescription(name = "group", description = "Manage the server groups")
public final class GroupCommand extends Command {

    @Override
    public void prepare() {
        registerSubCommand(SubCommand.lambda("create", this::handleCreateSub, true, subCommand -> {
            subCommand.addParameter(new StringParameter("name", false))
                    .addParameter(new StringParameter("templates", true, true));
        }));

        registerSubCommand(SubCommand.lambda("edit", this::handleEditSub, true, subCommand -> {
            subCommand.addParameter(new ServerGroupParameter("group", false))
                    .addParameter(new StringEnumParameter("action", false, "add", "remove"))
                    .addParameter(new StringParameter("template", false, true));
        }));

        registerSubCommand(SubCommand.lambda("remove", this::handleRemoveSub, true, subCommand -> {
            subCommand.addParameter(new ServerGroupParameter("group", false));
        }));

        registerSubCommand(SubCommand.lambda("info", this::handleInfoSub, true, subCommand -> {
            subCommand.addParameter(new ServerGroupParameter("group", false));
        }));

        registerSubCommand(SubCommand.lambda("list", this::handleListSub, true));
    }

    @Override
    public boolean run(CommandSender sender, CommandContext ctx) {
        return true;
    }

    public boolean handleCreateSub(CommandSender sender, CommandContext ctx) {
        String name = ctx.argString("name");
        if (PocketCloud.instance().serverGroups().check(name)) {
            sender.warn("A server group with that name already exists.");
            return true;
        }

        String providedTemplates = ctx.argString("templates");
        List<ITemplate> templates = new ArrayList<>();
        if (providedTemplates != null) {
            for (String template : providedTemplates.split(" ")) {
                if (PocketCloud.instance().templates().check(template)) {
                    templates.add(PocketCloud.instance().templates().get(template).get());
                }
            }
        }

        PocketCloud.instance().serverGroups().create(ServerGroupBuilder.create().name(name).templates(templates.toArray(ITemplate[]::new)));
        return true;
    }

    public boolean handleEditSub(CommandSender sender, CommandContext ctx) {
        ServerGroup group = ctx.arg("group", ServerGroup.class);
        //TODO
        return true;
    }

    public boolean handleRemoveSub(CommandSender sender, CommandContext ctx) {
        ServerGroup group = ctx.arg("group", ServerGroup.class);
        PocketCloud.instance().serverGroups().remove(group);
        return true;
    }

    public boolean handleInfoSub(CommandSender sender, CommandContext ctx) {
        ServerGroup group = ctx.arg("group", ServerGroup.class);
        sender.info("Name§8: §b{}", group.name());
        sender.info("Templates§8: §b{}", String.join("§8, §b", group.templates()));
        return true;
    }

    public boolean handleListSub(CommandSender sender, CommandContext ctx) {
        Collection<IServerGroup> groups = PocketCloud.instance().serverGroups().getAll();
        sender.info("ServerGroups §8(§b{}§8)§r:", groups.size());
        if (groups.isEmpty()) sender.info("§cNo server groups found.");
        for (IServerGroup group : groups) {
            sender.info("Name§8: §b{} §8| §rTemplates§8: §b{}", group.name(), String.join("§8, §b", group.templates()));
        }
        return true;
    }
}